sled-agent/src/nexus.rs

```diff
@@ -39,27 +39,57 @@ pub struct LazyNexusClient {
 }
 
 impl LazyNexusClient {
-    pub fn new(log: Logger, addr: Ipv6Addr) -> Result<Self, ResolveError> {
+    pub fn new_from_subnet(
+        log: Logger,
+        addr: Ipv6Addr,
+    ) -> Result<Self, ResolveError> {
         Ok(Self {
             inner: Arc::new(Inner {
                 log: log.clone(),
                 resolver: Resolver::new_from_ip(
                     log.new(o!("component" => "DnsResolver")),
                     addr,
                 )?,
+                address: Mutex::new(None),
+            }),
+        })
+    }
+
+    pub fn new_from_dns(
+        log: Logger,
+        dns_addrs: Vec<SocketAddr>,
+    ) -> Result<Self, ResolveError> {
+        Ok(Self {
+            inner: Arc::new(Inner {
+                log: log.clone(),
+                resolver: Resolver::new_from_addrs(
+                    log.new(o!("component" => "DnsResolver")),
+                    dns_addrs,
+                )?,
+                address: Mutex::new(None),
             }),
         })
     }
 
     pub async fn get_ip(&self) -> Result<Ipv6Addr, ResolveError> {
-        self.inner.resolver.lookup_ipv6(ServiceName::Nexus).await
+        self.get_addr().await.map(|addr| *addr.ip())
+    }
+
+    pub async fn get_addr(&self) -> Result<SocketAddrV6, ResolveError> {
+        if let Some(addr) = self.inner.address.lock().unwrap().as_ref() {
+            return Ok(*addr);
+        }
+        let addr =
+            self.inner.resolver.lookup_socket_v6(ServiceName::Nexus).await?;
+        *self.inner.address.lock().unwrap() = Some(addr);
+        Ok(addr)
```

Comment:
If I'm reading this right, this has the behavior that if several clients come into `get_addr()` concurrently, they will all try to look up the socket, and any that succeed will wind up storing whatever they find in here.  The last one will "win" forever -- we'll never look it up again.  That doesn't seem great.  We'll wind up hitting the DNS servers harder than we need to with a bunch of concurrent requests (which isn't a big deal right now) and we won't ever learn about any change in DNS.

I haven't taken a look at the surrounding code yet.  But are we not using a Progenitor-generated Oxide client?  I know we can give reqwest a DNS Resolver.  Is that not sufficient here?

Reply:
> I haven't taken a look at the surrounding code yet. But are we not using a Progenitor-generated Oxide client? I know we can give reqwest a DNS Resolver. Is that not sufficient here?

I've spent a while digging into this, implementing https://docs.rs/reqwest/latest/reqwest/dns/trait.Resolve.html for our `internal_dns` Resolver type, and I've come to the following conclusions:

- **Using the DNS resolver with progenitor / reqwest ignores the supplied port**. For reference, see: https://docs.rs/reqwest/0.11.18/reqwest/struct.ClientBuilder.html#method.resolve , but I also tested this manually. This sucks for testing, since we will need to know / commit to the port when the client is being constructed, and manually stick it in the URL. Granted, the ports **will** be stable for services in prod, so maybe we figure out a different workaround here when the system's under test.
- **Using the DNS resolver with progenitor / reqwest appears to query the DNS server on every request**. I tested this manually, but observed that there is no caching happening, and each individual client -> server request also hits the resolver. Perhaps this is fine, but I think it means that the responsibility for caching will need to live in the internal DNS resolver itself.

However, that being said, it *does* seem like using the [dns_resolver](https://docs.rs/reqwest/0.11.18/reqwest/struct.ClientBuilder.html#method.dns_resolver) method when constructing a Client, before it's passed to a progenitor client, does allow us to cope with the IP address of services shifting over time (and also, the DNS servers going up and down).

Reply:
To wrap this up: I've changed the name of this PR, and shifted the focus to "deprecate LazyNexusClient in favor of supplying the DNS resolver to the progenitor client (through the reqwest::Client)".